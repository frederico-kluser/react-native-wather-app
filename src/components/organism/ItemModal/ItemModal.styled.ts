import styled from 'styled-components/native';
import {deviceHeight, deviceWidth} from '../../../helpers/dimentions';

type Overflow = 'hidden' | 'scroll' | 'visible';
type Position = 'absolute' | 'relative';

interface viewStyleObjInterface {
  overflow: Overflow;
  position: Position;
}

const viewStyleObj: viewStyleObjInterface = {
  overflow: 'hidden',
  position: 'absolute',
};

export const viewStyle = (
  left: any,
  top: any,
  width: any,
  height: any,
  borderRadius: any,
) => ({
  backgroundColor: '#fff',
  borderRadius,
  height,
  left,
  top,
  width,
  zIndex: 1,
  ...viewStyleObj,
});

export const Container = styled.View`
  padding: 16px;
  padding-top: 64px;
  width: ${deviceWidth}px;
  height: ${deviceHeight}px;
`;

export const TitleCity = styled.Text`
  font-size: 30px;
  color: black;
`;

export const TitleTemperature = styled.Text`
  font-size: 96px;
  color: black;
`;

export const TitleIcon = styled.Image`
  height: 256px;
  position: absolute;
  right: -64px;
  top: 0px;
  width: 256px;
`;

export const ContentContainer = styled.View`
  position: absolute;
  bottom: 15%;
  left: 16px;
  height: auto;
`;
