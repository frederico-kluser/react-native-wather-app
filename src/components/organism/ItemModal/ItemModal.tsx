import React, {useState, useEffect} from 'react';
import {Animated} from 'react-native';
import kluserMotion, {animationTime} from '../../../helpers/animation';
import {deviceHeight, deviceWidth, itemSize} from '../../../helpers/dimentions';
import {weatherIcons} from '../../../styles/icons';
import {PAGE_PADDING} from '../../../styles/theme';
import HorizontalCarousel from '../../molecule/HorizontalCarousel/HorizontalCarousel';
import VerticalList from '../../molecule/VerticalList/VerticalList';
import Button, {
  BG,
  Container,
  ContentContainer,
  TitleCity,
  TitleIcon,
  TitleTemperature,
  viewStyle,
} from './ItemModal.styled';

const ItemModal = ({
  index,
  city,
  status,
  setSelectedItem,
  scrollY,
  temperature,
}: any) => {
  const distanceFromTop = Math.ceil((index + 1) / 2);
  const sizeMultiplier = distanceFromTop - 1;
  const marginMultiplier = sizeMultiplier
    ? PAGE_PADDING * 2 * sizeMultiplier
    : 0;
  const topSize = itemSize * sizeMultiplier + marginMultiplier + PAGE_PADDING;

  const leftSize = index % 2 === 0 ? 16 : deviceWidth - (itemSize + 16);

  const [style, setStyle] = useState({
    borderRadius: kluserMotion(16, 0),
    top: kluserMotion(topSize, scrollY),
    left: kluserMotion(leftSize, 0),
    width: kluserMotion(itemSize, deviceWidth),
    height: kluserMotion(itemSize, deviceHeight),
  });

  useEffect(() => {
    style.borderRadius.start();
    style.top.start();
    style.left.start();
    style.width.start();
    style.height.start();
  }, [style]);

  return (
    <Animated.View
      style={viewStyle(
        style.left.animation,
        style.top.animation,
        style.width.animation,
        style.height.animation,
        style.borderRadius.animation,
      )}>
      <Button
        index={index}
        onPress={() => {
          setStyle({
            borderRadius: style.borderRadius.reverse(),
            top: style.top.reverse(),
            left: style.left.reverse(),
            width: style.width.reverse(),
            height: style.height.reverse(),
          });
          setTimeout(() => {
            setSelectedItem(-1);
          }, animationTime);
        }}
      />
      <Container>
        <TitleIcon source={weatherIcons[status]} />
        <TitleCity>{city}</TitleCity>
        <TitleTemperature>{temperature}Â°</TitleTemperature>
        <ContentContainer>
          <HorizontalCarousel />
          <VerticalList />
        </ContentContainer>
      </Container>
    </Animated.View>
  );
};

export default ItemModal;
