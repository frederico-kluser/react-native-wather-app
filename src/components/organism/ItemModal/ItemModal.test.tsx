import React from 'react';
import renderer from 'react-test-renderer';
import ItemModal from './ItemModal';

jest.useFakeTimers();

test('renders correctly', () => {
  const tree = renderer.create(<ItemModal />).toJSON();
  expect(tree).toMatchSnapshot();
});
