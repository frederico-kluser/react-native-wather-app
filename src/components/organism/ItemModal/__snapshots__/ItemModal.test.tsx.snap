// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  collapsable={false}
  style={
    Object {
      "backgroundColor": "#fff",
      "borderRadius": 16,
      "height": 343,
      "left": 391,
      "overflow": "hidden",
      "position": "absolute",
      "top": NaN,
      "width": 343,
      "zIndex": 1,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "left": 16,
        "opacity": 1,
        "position": "absolute",
        "top": 16,
        "zIndex": 2,
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/heart.png",
        }
      }
      style={
        Array [
          Object {
            "height": 32,
            "width": 32,
          },
        ]
      }
    />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
        "position": "absolute",
        "right": 16,
        "top": 16,
        "zIndex": 2,
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/close2.png",
        }
      }
      style={
        Array [
          Object {
            "height": 32,
            "width": 32,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "height": 1334,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 64,
          "width": 750,
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/sun.png",
        }
      }
      style={
        Array [
          Object {
            "height": 256,
            "position": "absolute",
            "right": -64,
            "top": 0,
            "width": 256,
          },
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontSize": 30,
          },
        ]
      }
    >
      New York
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontSize": 96,
          },
        ]
      }
    >
      6
      °
    </Text>
    <View
      style={
        Array [
          Object {
            "top": "8%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "marginBottom": "6%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 13,
                "fontWeight": "500",
                "textTransform": "capitalize",
              },
            ]
          }
        >
          Terça
        </Text>
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/rain.png",
            }
          }
          style={
            Array [
              Object {
                "height": 32,
                "left": "47%",
                "position": "absolute",
                "top": -8,
                "width": 32,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "red",
                "fontSize": 13,
                "fontWeight": "bold",
                "left": "80%",
                "position": "absolute",
              },
            ]
          }
        >
          0
          °
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "blue",
                "fontSize": 12,
                "fontWeight": "bold",
                "position": "absolute",
                "right": "0%",
              },
            ]
          }
        >
          6
          °
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "marginBottom": "6%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 13,
                "fontWeight": "500",
                "textTransform": "capitalize",
              },
            ]
          }
        >
          Quarta
        </Text>
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/sun.png",
            }
          }
          style={
            Array [
              Object {
                "height": 32,
                "left": "47%",
                "position": "absolute",
                "top": -8,
                "width": 32,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "red",
                "fontSize": 13,
                "fontWeight": "bold",
                "left": "80%",
                "position": "absolute",
              },
            ]
          }
        >
          0
          °
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "blue",
                "fontSize": 12,
                "fontWeight": "bold",
                "position": "absolute",
                "right": "0%",
              },
            ]
          }
        >
          6
          °
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "marginBottom": "6%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 13,
                "fontWeight": "500",
                "textTransform": "capitalize",
              },
            ]
          }
        >
          Quinta
        </Text>
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/cloud.png",
            }
          }
          style={
            Array [
              Object {
                "height": 32,
                "left": "47%",
                "position": "absolute",
                "top": -8,
                "width": 32,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "red",
                "fontSize": 13,
                "fontWeight": "bold",
                "left": "80%",
                "position": "absolute",
              },
            ]
          }
        >
          1
          °
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "blue",
                "fontSize": 12,
                "fontWeight": "bold",
                "position": "absolute",
                "right": "0%",
              },
            ]
          }
        >
          8
          °
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "marginBottom": "6%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 13,
                "fontWeight": "500",
                "textTransform": "capitalize",
              },
            ]
          }
        >
          Sexta
        </Text>
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/sun.png",
            }
          }
          style={
            Array [
              Object {
                "height": 32,
                "left": "47%",
                "position": "absolute",
                "top": -8,
                "width": 32,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "red",
                "fontSize": 13,
                "fontWeight": "bold",
                "left": "80%",
                "position": "absolute",
              },
            ]
          }
        >
          0
          °
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "blue",
                "fontSize": 12,
                "fontWeight": "bold",
                "position": "absolute",
                "right": "0%",
              },
            ]
          }
        >
          7
          °
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "marginBottom": "6%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 13,
                "fontWeight": "500",
                "textTransform": "capitalize",
              },
            ]
          }
        >
          Sábado
        </Text>
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/cloud.png",
            }
          }
          style={
            Array [
              Object {
                "height": 32,
                "left": "47%",
                "position": "absolute",
                "top": -8,
                "width": 32,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "red",
                "fontSize": 13,
                "fontWeight": "bold",
                "left": "80%",
                "position": "absolute",
              },
            ]
          }
        >
          4
          °
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "blue",
                "fontSize": 12,
                "fontWeight": "bold",
                "position": "absolute",
                "right": "0%",
              },
            ]
          }
        >
          13
          °
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "marginBottom": "6%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 13,
                "fontWeight": "500",
                "textTransform": "capitalize",
              },
            ]
          }
        >
          Domingo
        </Text>
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/sunSnow.png",
            }
          }
          style={
            Array [
              Object {
                "height": 32,
                "left": "47%",
                "position": "absolute",
                "top": -8,
                "width": 32,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "red",
                "fontSize": 13,
                "fontWeight": "bold",
                "left": "80%",
                "position": "absolute",
              },
            ]
          }
        >
          -3
          °
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "blue",
                "fontSize": 12,
                "fontWeight": "bold",
                "position": "absolute",
                "right": "0%",
              },
            ]
          }
        >
          0
          °
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "marginBottom": "6%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 13,
                "fontWeight": "500",
                "textTransform": "capitalize",
              },
            ]
          }
        >
          Segunda
        </Text>
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/cloudWind.png",
            }
          }
          style={
            Array [
              Object {
                "height": 32,
                "left": "47%",
                "position": "absolute",
                "top": -8,
                "width": 32,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "red",
                "fontSize": 13,
                "fontWeight": "bold",
                "left": "80%",
                "position": "absolute",
              },
            ]
          }
        >
          -5
          °
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "blue",
                "fontSize": 12,
                "fontWeight": "bold",
                "position": "absolute",
                "right": "0%",
              },
            ]
          }
        >
          0
          °
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": "11%",
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/assets/trash.png",
          }
        }
        style={
          Array [
            Object {
              "bottom": 20,
              "left": "50%",
              "position": "absolute",
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "left": 16,
              "textAlign": "center",
            },
          ]
        }
      >
        Deletar Item
      </Text>
    </View>
  </View>
</View>
`;
